### ISSUE TITLE
Change Request: Eslintcache relative

### ISSUE THREAD

**Issue by brenelz** [eslint/eslint#16493]
### ESLint version

v8.26.0

### What problem do you want to solve?

I am currently trying to use the .eslintcache file in our jenkins ci.

Is there anyway the cache can use relative file paths? In our instance each branch is in a seperate dir but I want to reuse the cache.

### What do you think is the correct solution?

In order to move the .eslintcache around it would be great if the cache filePaths were relative

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by mdjermanovic** [comment]
Hi @brenelz, thanks for the issue!

I think this is generally doable. We'd need to discuss details, like whether this needs a new option and what should be the base path (directory to which the file paths would be relative; would it be the current working directory, or location of the cache file, or something else). This is a non-trivial change in the core, so I think it should go through the [RFC process](https://github.com/eslint/rfcs), but let's hear more opinions from the team first.

**Comment by brenelz** [comment]
If its a non-trivial change its probably not worth it. Seems like no one has really asked for it besides me. We were just trying to use the cache in our ci which I think we are unable to do currently.

**Comment by brenelz** [comment]
Another option very specific to my use case would be if we could not include the path in hashOfConfig if that makes sense

**Comment by nzakas** [comment]
Yeah, we’d need an RFC for this as it has the potential to break current cache file usage.

**Comment by CSchulz** [comment]
Who can initiate this rfc request? We are also struggling with caching.

Why not using a hybrid scenario for the paths? Cutting the absolute path to a relative path and always store the relative path for the future.
As long as nobody moves the workspace around there is no breaking change for the cache itself.

**Comment by nzakas** [comment]
Anyone can initiate an RFC. Some people check in their cache file, so any change to the cache format is a breaking change. We can't know for sure where people are storing cache files or how else they might be used.

**Comment by CSchulz** [comment]
I can't believe anyone is using absolute paths if people want to check in their cache files.
Otherwise to prevent some breaking changes, can we add a third strategy for the cache using relative paths?

**Comment by nzakas** [comment]
@CSchulz you can feel free to propose whatever you want. As already stated, we would just need an RFC to consider it.

**Comment by aaronabf** [comment]
> If its a non-trivial change its probably not worth it. Seems like no one has really asked for it besides me. We were just trying to use the cache in our ci which I think we are unable to do currently.

I second this feature! I had opened an issue for this as well [#16741](https://github.com/eslint/eslint/issues/16741).

**Comment by herzaso** [comment]
+1.
I wanted to use caching and check in the cache file so that other developers in my organization (and the CI) would use that cache, but it is currently not feasible.
The biggest problem is the CI - since we use docker to run the lint, it can't maintain the `.eslintcache` file so we have to check it in, but then I don't think that the CI should write anything to my repo

**Comment by majo44** [comment]
+1 on our big monorepo (500 packages) we also want to share the `.eslintcache` especially to save CI time.

**Comment by Rec0iL99** [comment]
WIP. Don't close. @CSchulz has opened a [RFC](https://github.com/eslint/rfcs/pull/106).

**Comment by cs6cs6** [comment]
I would like to see this proceed.

**Comment by brenelz** [comment]
Personally I don't have a specific need for this anymore, but I think it would still be nice

**Comment by cs6cs6** [comment]
I do. I'm prototyping it, and if it works I'll try to submit this back. 
--------------

lib/options.js: Add the option to specify a non default estlintcache directory structure, around line 313:
   option: cachefolderstorage
   type: String
   enum: ["full", "relative"],
   default: "full"
   description: "Indicates how an eslintcache will internally store directories. "relative" allows for the same cache to be used at different directory paths. "full" is the legacy, default behavior where the full file path is encoded in the eslintcache.

 

conf/default-cli-options.js: Add the name "cachefolderstorage" with a default of "full" around line 25

 

eslint/eslint-hepers.js: Modify processOptions function around line 662 to have the field "cachefolderstorage" with a default of "full"

 

lib/cli-engine/lint-result-cache.js: add a debug printout to state what kind of directory struture is used in the LintResultCache (print out cachefolderstorage value)

 

lib/eslibt/flat-eslint.js: In function "lintFiles" which starts around line 733, the list of file directory paths is created from the pattern passed in on the cli. This 
list of files is ultimately what gets written to the eslintcache. Those file names are full path names. 
   The easiest change here with the least side effects is likely just to change the way the result object is stored. On Line 871 the code looks like this:
                        if (lintResultCache) {
                            lintResultCache.setCachedLintResults(filePath, config, result);
                        }
   It would likely be best to look at the config option cachefolderstorage here, and see if filePath should be stored as a relative or a full path.

 

 

Some unit tests will definitely need to be added to the flat-eslint tests, and the option-parser tests.

**Comment by CSchulz** [comment]
@cs6cs6 I am out of time at the moment, but I would suggest to take over my rfc request or we work together?

**Comment by nzakas** [comment]
This isn't stale, we have an RFC open.

**Comment by nzakas** [comment]
We have an RFC that hasn't been updated in a bit. Will ping the author.

**Comment by nzakas** [comment]
We are waiting on the RFC to be picked up.

**Comment by snitin315** [comment]
Assigned it to the RFC author, to avoid stale-bot messages.

**Comment by mdjermanovic** [comment]
I'll submit a new RFC.

**Comment by aaronabf** [comment]
It's crazy this is still open.
